### PR TITLE
Allow passing locks to C code.

### PR DESCRIPTION
**Pull request by stephenneuendorffer**

### CODE REVIEW

File: test/unit_tests/chess_compiler_tests/01_precompiled_core_function/kernel.cc [hunk 1]
```diff
@@ -10,13 +10,15 @@
 
 #include "kernel.h"
 
-void func(int32_t *a, int32_t *b)
+void func(int32_t *a, int32_t *b, int64_t lock)
```
Comment (keryell) [inline test/unit_tests/chess_compiler_tests/01_precompiled_core_function/kernel.cc:13]: Bad for the carbon footprint to have 64 bits here?

File: test/unit_tests/chess_compiler_tests/01_precompiled_core_function/aie.mlir [hunk 2]
```diff
@@ -22,12 +22,12 @@ module @test_chesss_01_precompiled_core_function {
   %lock13_3 = AIE.lock(%tile13, 3) { sym_name = "input_lock" }
   %lock13_5 = AIE.lock(%tile13, 5) { sym_name = "output_lock" }
 
-  func.func private @func(%A: memref<256xi32>, %B: memref<256xi32>) -> ()
+  func.func private @func(%A: memref<256xi32>, %B: memref<256xi32>, %lock: index) -> ()
```
Comment (keryell) [inline test/unit_tests/chess_compiler_tests/01_precompiled_core_function/aie.mlir:25]: Perhaps `index` could be a more precise type depending on the architecture?
Or at least lowered to a more precise concrete type before calling the real AIE kernel?

File: test/unit_tests/chess_compiler_tests/01_precompiled_core_function/kernel.cc (hunk 1 above)
Comment (stephenneuendorffer) [reply test/unit_tests/chess_compiler_tests/01_precompiled_core_function/kernel.cc:13]: Yes.. this is the main reason why this hasn't been checked in.  Today, it's relatively easy to lower from locks to a constant 'index' type in MLIR, but the size of the index type is something 'implementation defined' (by default 64 bits).   This really should be cleaner in the lowering to result in something more convenient.

**Comment by keryell** [comment]
To be merged?

**Pull request closed without merging**